toolbox/relnotes/main.go

```diff
@@ -0,0 +1,757 @@
+	generateMDFile(client, releaseInfo.releaseTag, prFileName)
+
+	if *htmlizeMD {
+		// HTML-ize markdown file
+		// Make users and PRs linkable
+		// Also, expand anchors (needed for email announce())
+		projectGithubURL := fmt.Sprintf("https://github.com/%s/%s", *owner, *repo)
+		_, err := u.Shell("sed", "-i", "-e", "s,#\\([0-9]\\{5\\,\\}\\),[#\\1]("+projectGithubURL+"/pull/\\1),g",
+			"-e", "s,\\(#v[0-9]\\{3\\}-\\),"+projectGithubURL+"/blob/master/CHANGELOG.md\\1,g",
+			"-e", "s,@\\([a-zA-Z0-9-]*\\),[@\\1](https://github.com/\\1),g", *mdFileName)
+
+		if err != nil {
+			log.Printf("failed to htmlize markdown file: %s", err)
+			os.Exit(1)
+		}
+	}
+
+	if *preview && *owner == "kubernetes" && *repo == "kubernetes" {
+		// If in preview mode, get the current CI job status
+		// We do this after htmlizing because we don't want to update the
+		// issues in the block of this section
+		//
+		// NOTE: this function is Kubernetes-specified and runs the find_green_build script under
+		// kubernetes/release. Make sure you have the dependencies installed for find_green_build
+		// before running this function.
+		err := getCIJobStatus(*mdFileName, *branch, *htmlizeMD)
+		if err != nil {
+			log.Printf("failed to get CI status: %s", err)
+			os.Exit(1)
+		}
+	}
+
+	if *htmlFileName != "" {
+		// If HTML file name is given, generate HTML release note
+		err := createHTMLNote(*htmlFileName, *mdFileName)
+		if err != nil {
+			log.Printf("failed to generate HTML release note: %s", err)
+			os.Exit(1)
+		}
+	}
+
+	if !*quiet {
+		// If --quiet flag is not specified, print the markdown release note to stdout
+		log.Printf("Displaying the markdown release note to stdout...")
+		dat, err := ioutil.ReadFile(*mdFileName)
+		if err != nil {
+			log.Printf("failed to read markdown release note: %s", err)
+			os.Exit(1)
+		}
+		fmt.Print(string(dat))
+	}
+
+	log.Printf("Successfully generated release note. Total running time: %s", time.Now().Round(time.Second).Sub(startingTime).String())
+
+	return
+}
+
+func readGithubToken(filename string) string {
+	dat, err := ioutil.ReadFile(filename)
+	if err != nil {
+		log.Printf("failed to get Github token from file %s: %s", filename, err)
+		os.Exit(1)
+	}
+	return strings.TrimSpace(string(dat))
+}
+
+func gatherReleaseInfo(client *github.Client, branchRange string) *ReleaseInfo {
+	var info ReleaseInfo
+	log.Printf("Gathering release commits from Github...")
+	// Get release related commits on the release branch within release range
+	releaseCommits, startTag, releaseTag, err := getReleaseCommits(client, *owner, *repo, *branch, branchRange)
+	if err != nil {
+		log.Printf("failed to get release commits for %s: %s", branchRange, err)
+		os.Exit(1)
+	}
+	info.startTag = startTag
+	info.releaseTag = releaseTag
+
+	// Parse release related PR ids from the release commits
+	commitPRs, err := parsePRFromCommit(releaseCommits)
+	if err != nil {
+		log.Printf("failed to parse release commits: %s", err)
+		os.Exit(1)
```

Comment:
You should only use os.Exit() in main(), use return (with/without err)

Reply:
Done